prov/usnic/src/usdf_pep.c

```diff
@@ -337,6 +337,104 @@ usdf_pep_listen(struct fid_pep *fpep)
 	return 0;
 }
 
+/* Register as a callback triggered by the socket becoming writeable. Write as
+ * much data as can be written in a single write, and keep track of how much
+ * data is left. If the data is not fully written, it will finish getting
+ * written in another iteration of the progression.
+ */
+static int usdf_pep_reject_async(void *vreq)
+{
+	struct usdf_connreq *crp;
+	struct usdf_eq *eq;
+	int ret;
+
+	crp = vreq;
+	eq = crp->cr_pep->pep_eq;
+
+	do {
+		ret = write(crp->cr_sockfd, crp->cr_ptr, crp->cr_resid);
+	} while ((ret < 0) && (errno == EINTR || errno == EAGAIN));
```

Comment:
> IIRC, when O_NONBLOCK is set and you exceed the file's/socket's write buffer, you could get back either -1 with EAGAIN/EWOULDBLOCK or you could get back a return value of 0. I think the former behavior is more common.

I think if the send/write call returns `0`, then the socket has been closed. That means the conditional below the loop should be `if (ret <= 0)` instead. 

So I think the update for this should look more like: 

```c
do {
    ret = write(crp->cr_sockfd, crp->cr_ptr, crp->cr_resid);
} while ((ret < 0) && (errno == EINTR));

if ((ret <= 0) && (errno != EAGAIN)) {
    USDF_DBG_SYS(EP_CTRL, "write failed: %s\n",
 				 strerror(errno));
    usdf_cm_msg_connreq_failed(crp, -errno);
    return -errno;
}
```

Agreed?

The problem with this is that the progression thread is setup to exit if any progress call fails. See [here](https://github.com/bturrubiates/libfabric/blob/eae35bf5dfe718e24d23314aebb8e04e4196d992/prov/usnic/src/usdf_progress.c#L125). So getting `EAGAIN` isn't an unrecoverable error because it will become writable on the next `EPOLLOUT` event and call the callback again. But if the send call errored with anything other than `EAGAIN` or `EINTR`, then this will probably kill the progression thread.